models/src/main/protobuf/CasperMessage.proto

```diff
@@ -119,6 +119,8 @@ message BlockMetadataInternal {
   int64 blockNum                        = 6;
   int32 seqNum                          = 7;
   bool invalid                          = 8; // whether the block was marked as invalid
+  bool directlyFinalized                = 9; // whether the block has been last finalized block (LFB)
+  bool finalized                        = 10;// whether the block is finalized
```

Comment:
Why do we need two fields for that?
What is the goal of `directlyFinalized` trying to solve?

Reply:
`directlyFinalized` means it has been output by finalizer, which means it has been last finalized block at some point. And LFB is used to constrain merging scope, we need at least two (well, exactly two I guess) merging scopes to reliably check if deploy cannot be rejected.

Reply:
:) I still do not quite understand the scenario. Maybe I don't understand from the point of choosing merging scope by LFB from the beginning which makes me understand the `directlyFinalized` difficulty.

`All the finalized blocks should be last finalized block at some poing.` This is wrong ?

Reply:
`All the finalized blocks should be last finalized block at some poing` yes this is wrong. E.g. LFB1 can be at height 1 and next one, LFB2 - at height 5. So all blocks that are ancestors of LFB2 should be finalized as well. But they are not found by Finalizer according to clique oracle rule.

In this scenario between LFB1 and LFB2 there will be multiple merge blocks, all of them using LFB1 as a base for merge (state that cannot be changed).